message/validation/alan/consensus_validation.go

```diff
@@ -0,0 +1,427 @@
+			e := ErrNonDecidedWithMultipleSigners
+			e.got = len(signers)
+			return e
+		}
+
+		if uint64(len(signers)) < quorumSize {
+			e := ErrDecidedNotEnoughSigners
+			e.want = quorumSize
+			e.got = len(signers)
+			return e
+		}
+	}
+
+	if len(signedSSVMessage.FullData) != 0 {
+		if consensusMessage.MsgType == specqbft.PrepareMsgType ||
+			consensusMessage.MsgType == specqbft.CommitMsgType && len(signedSSVMessage.GetOperatorIDs()) == 1 {
+			return ErrPrepareOrCommitWithFullData
+		}
+
+		hashedFullData, err := specqbft.HashDataRoot(signedSSVMessage.FullData)
+		if err != nil {
+			e := ErrFullDataHash
+			e.innerErr = err
+			return e
+		}
+
+		if hashedFullData != consensusMessage.Root {
+			return ErrInvalidHash
+		}
+	}
+
+	if !mv.validConsensusMsgType(consensusMessage.MsgType) {
+		return ErrUnknownQBFTMessageType
+	}
+
+	if consensusMessage.Round == specqbft.NoRound {
+		e := ErrInvalidRound
+		e.got = specqbft.NoRound
+		return e
+	}
+
+	if !bytes.Equal(consensusMessage.Identifier, signedSSVMessage.GetSSVMessage().MsgID[:]) {
+		e := ErrMismatchedIdentifier
+		e.want = hex.EncodeToString(signedSSVMessage.GetSSVMessage().MsgID[:])
+		e.got = hex.EncodeToString(consensusMessage.Identifier)
+		return e
+	}
+
+	if err := mv.validateJustifications(consensusMessage); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (mv *messageValidator) validateQBFTLogic(
+	signedSSVMessage *spectypes.SignedSSVMessage,
+	consensusMessage *specqbft.Message,
+	committee []spectypes.OperatorID,
+	receivedAt time.Time,
+	state *consensusState,
+) error {
+	if consensusMessage.MsgType == specqbft.ProposalMsgType {
+		leader := mv.roundRobinProposer(consensusMessage.Height, consensusMessage.Round, committee)
+		if signedSSVMessage.GetOperatorIDs()[0] != leader {
+			err := ErrSignerNotLeader
+			err.got = signedSSVMessage.GetOperatorIDs()[0]
+			err.want = leader
+			return err
+		}
+	}
+
+	msgSlot := phase0.Slot(consensusMessage.Height)
+	for _, signer := range signedSSVMessage.GetOperatorIDs() {
+		signerState := state.GetSignerState(signer)
+		if signerState == nil {
+			continue
+		}
+
+		// It should be checked after ErrNonDecidedWithMultipleSigners
+		signerCount := len(signedSSVMessage.GetOperatorIDs())
+		if signerCount > 1 {
+			if _, ok := signerState.SeenDecidedLengths[signerCount]; ok {
+				return ErrDecidedWithSameNumberOfSigners
+			}
+		}
+
+		if msgSlot == signerState.Slot && consensusMessage.Round == signerState.Round {
+			if len(signedSSVMessage.FullData) != 0 && signerState.ProposalData != nil && !bytes.Equal(signerState.ProposalData, signedSSVMessage.FullData) {
+				return ErrDuplicatedProposalWithDifferentData
+			}
+
+			limits := maxMessageCounts(len(committee))
+			if err := signerState.MessageCounts.ValidateConsensusMessage(signedSSVMessage, consensusMessage, limits); err != nil {
+				return err
+			}
+		}
+	}
+
+	return mv.roundBelongsToAllowedSpread(signedSSVMessage, consensusMessage, receivedAt)
```

Comment:
I've added the `PeerAlreadyAdvancedRound` rule after this one. Is it already added somewhere else?

Reply:
Could you please explain why we should check only the round but not the slot?

Reply:
Regarding the "already advanced", for the same duty instance, it's fine for us to ignore old messages of it (like previous rounds). But, the slot rule limits the number of ongoing duties. I don't know if the committee runner may have 2 consensus running at the same time. So, better not to add now

Reply:
@MatheusFranco99 and I decided to put this question on hold to discuss the possibility of having several duties for a sender/role pair